iohelper/iohelper.go

```diff
@@ -0,0 +1,95 @@
+package iohelper
+
+/*
+ * This file contains structs and functions related to interacting with files
+ * and directories.
+ */
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/greenplum-db/gp-common-go-libs/gplog"
+	"github.com/greenplum-db/gp-common-go-libs/operating"
+	"github.com/pkg/errors"
+)
+
+/*
+ * Generic file/directory manipulation functions that use the function pointers
+ * and types from the operating package, for easier test mocking around file IO.
+ */
+
+func OpenFileForReading(filename string) (operating.ReadCloserAt, error) {
+	fileHandle, err := operating.System.OpenFileRead(filename, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, errors.Errorf("Unable to open file for reading: %s", err)
+	}
+	return fileHandle, nil
+}
+
+func MustOpenFileForReading(filename string) operating.ReadCloserAt {
+	fileHandle, err := OpenFileForReading(filename)
+	gplog.FatalOnError(err)
+	return fileHandle
+}
+
+func OpenFileForWriting(filename string, allowAppend ...bool) (io.WriteCloser, error) {
+	flags := os.O_CREATE | os.O_WRONLY
+	if len(allowAppend) == 1 && allowAppend[0] {
+		flags = os.O_APPEND | flags
+	}
+	fileHandle, err := operating.System.OpenFileWrite(filename, flags, 0644)
+	if err != nil {
+		return nil, errors.Errorf("Unable to create or open file for writing: %s", err)
+	}
+	return fileHandle, nil
+}
+
+func MustOpenFileForWriting(filename string, allowAppend ...bool) io.WriteCloser {
+	fileHandle, err := OpenFileForWriting(filename, allowAppend...)
+	gplog.FatalOnError(err)
+	return fileHandle
+}
+
+func FileExistsAndIsReadable(filename string) bool {
+	_, err := operating.System.Stat(filename)
+	if err == nil {
+		var fileHandle io.ReadCloser
+		fileHandle, err = OpenFileForReading(filename)
+		if fileHandle != nil {
+			fileHandle.Close()
+		}
+		if err == nil {
+			return true
+		}
+	}
+	return false
+}
+
+func ReadLinesFromFile(filename string) ([]string, error) {
+	fileHandle, err := OpenFileForReading(filename)
+	if err != nil {
+		return nil, err
+	}
+	return readLines(fileHandle), nil
+}
+
+func MustReadLinesFromFile(filename string) []string {
+	fileHandle := MustOpenFileForReading(filename)
+	return readLines(fileHandle)
+}
+
+// Read all lines from a file, stripping whitespace and ignoring empty lines
+func readLines(fileHandle operating.ReadCloserAt) []string {
```

Comment:
Do you see people using this from a ReadCloser, or would they usually just want to read from a filename? If the latter, then maybe we can combine them into one function.

Reply:
I know there was a reason I pulled this out instead of calling `ReadLinesFromFile` directly , but I don't recall what it was, so it must not have been that compelling.  I'll fold it back in.